wasm/src/language_service.rs

```diff
@@ -1,40 +1,194 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
+use std::{future::Future, path::PathBuf, pin::Pin, sync::Arc};
+
 use crate::{diagnostic::VSDiagnostic, serializable_type};
 use js_sys::JsString;
 use qsc::{self};
+use qsc_project::{EntryType, Manifest, ManifestDescriptor};
+use qsls::JSFileEntry;
 use rustc_hash::FxHashMap;
 use serde::{Deserialize, Serialize};
 use wasm_bindgen::prelude::*;
+use wasm_bindgen_futures::JsFuture;
 
 #[wasm_bindgen]
 pub struct LanguageService(qsls::LanguageService<'static>);
 
+/// This macro calls an async JS function, awaits it, and then applies a transformer function to it. Ultimately, it returns a function that accepts a String
+/// and returns a JS promise that is represented by a Rust future. (I know. Ouch. My head.)
+macro_rules! call_async_js_fn {
+    ($js_async_fn: ident, $transformer: expr) => {{
+        let $js_async_fn = to_js_function($js_async_fn.obj, stringify!($js_async_fn));
+
+        let $js_async_fn = move |input: String| {
+            let path = JsValue::from_str(&input);
+            let res: js_sys::Promise = $js_async_fn
+                .call1(&JsValue::NULL, &path)
+                .expect("callback should succeed")
+                .into();
+
+            let res: JsFuture = res.into();
+
+            let input = input.clone();
+            Box::pin(map_js_promise(res, move |x| $transformer(x, input.clone())))
+                as Pin<Box<dyn Future<Output = _> + 'static>>
+        };
+        $js_async_fn
+    }};
+}
+
+async fn map_js_promise<F, T>(res: JsFuture, func: F) -> T
+where
+    F: Fn(JsValue) -> T,
+{
+    let res = res.await.expect("js future shouldn't throw an exception");
```

Comment:
Doesn't have to happen in this PR but IMO it would be more appropriate to return the failure here and have the _caller_  do the `.expect()` if it wants to. Since this is a generic wrapper function it makes me nervous that it makes assumptions about the fallibility of the wrapped JS function.